docs/architecture.md

```diff
@@ -0,0 +1,185 @@
+
+The CVO has [some documentation](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md) about how to interact with it. Most importantly, the CNO will create everything in the `/manifests` folder.
+
+
+
+### Install payload
+
+The CVO tracks the "real" paths to built images, as those are dependent on build details, CI status, and offline clusters. It passes these to the CNO at runtime via environment variables.
+
+The key takeaway is this: **only reference images provided by the CVO**. You may not reference any other images, even if they're located at `quay.io/openshift`.
+
+### ClusterOperator Status
+
+All SLOs must publish their status to a CustomResource of type ClusterOperator. This is how operators report failing conditions to administrators, as well as blocking upgrades.
+
+There is code to manage generating this status - controllers just need to determine if they are degraded or not.
+
+
+## Controllers
+
+The individual controllers are all (mostly) independent. If they do communicate, it is generally through the apiserver. That means that each controller could theoretically be a separate process.
+
+Controllers all live in `./pkg/controller/`
+
+## Cluster Config Controller
+
+**Input:** `Network.config.openshift.io/v1 cluster` ([typedef](https://github.com/openshift/api/blob/master/config/v1/types_network.go))
+
+**Output:** `Network.operator.openshift.io/v1 cluster` ([typedef](https://github.com/openshift/api/blob/master/operator/v1/types_network.go))
+
+The Cluster Config controller reads the high-level configuration object, and applies it "downward" to the detailed network configuration (the poorly-named _Operator Configuration_). Any conflicting fields in the operator configuration will be overwritten.
+
+It partially generates the status of the `Network.config` object, but this object is a bit of an exception.
+
+## Operator Config Controller (Network Controller)
+
+**Input:** `Network.operator.openshift.io/v1 cluster` ([typedef](https://github.com/openshift/api/blob/master/operator/v1/types_network.go))
+
+**Output:** Most core networking components.
+
+This is the "main" controller, in that it is responsible for rendering the core networking components (openshift-sdn / OVNKubernetes / Kuryr, Multus, etc). It is broken down in to stages:
+
+1. **Validate** the supplied configuration
+2. **Fill** - Determine any unsupplied default values.
+3. **Check** - Compare against previously-applied configuration, to see if any unsafe changes are proposed
+4. **Bootstrap** - gather existing cluster state, and create any non-Kubernetes resources (i.e. OpenStack objects)
+5. **Render** - process template files in `/bindata` and generate Kubernetes objects
```

Comment:
maybe it is worth to add a subsection for the upgrade logic  (and the dual-stack conversion I've added) ?
 that make use of these 2 steps to retain one of the changes